singletons.go

```diff
@@ -0,0 +1,46 @@
+package lxr
+
+import "sync"
+
+// The goal of instances is to provide a way for multiple packages to use LXR without
+// instantiating multiple bytemaps in memory or having to share references
+
+var instanceMtx sync.Mutex
+var instances map[uint64]*LXRHash
+
+func init() {
+	instances = make(map[uint64]*LXRHash)
+}
+
+// Init provides access to shared instances of LXRHash without having to instantiate multiple bytemaps.
+// Two separate calls to Init(n) will result in a reference to the same object.
+// LXRHash will be instantiated with the package defaults of:
+// 	* Seed: 0xFAFAECECFAFAECEC
+// 	* Hash Size: 256
+// 	* Passes: 5
+func Init(bitsize uint64) *LXRHash {
+	if bitsize < 8 {
+		panic("bitsize must be at least 8")
+	}
+
+	instanceMtx.Lock()
+	defer instanceMtx.Unlock()
+
+	if instance, ok := instances[bitsize]; ok {
+		return instance
+	}
+
+	lxr := new(LXRHash)
+	lxr.Verbose(true)
+	lxr.Init(Seed, bitsize, HashSize, Passes)
+	instances[bitsize] = lxr
+	return lxr
+}
+
+// Release dereferences the shared instance inside this package, allowing the garbage collector to free the memory.
+// Please note that any existing references to the instance outside this package will keep it alive
+func Release(bitsize uint64) {
+	instanceMtx.Lock()
+	defer instanceMtx.Unlock()
+	delete(instances, bitsize)
+}

```

Comment:
The Release function doesn't take into account any other users of the instance of the LXRHash.  So a bit of code that "releases" the instance doesn't care about other code that might be using the instance.  If you then ask for a Singleton, you will get a new instance in addition to any code using the prior instance.

I'd suggest just getting rid of Release().  Or add reference counting so you don't actually delete the instance unless you have a release for every Singleton request. 

This still means the code can fail to release Singletons if it doesn't balance requests with releases.  For our purposes, I think just living with the Singleton if one is requested is the better path?

Reply:
> The Release function doesn't take into account any other users of the instance of the LXRHash. So a bit of code that "releases" the instance doesn't care about other code that might be using the instance. If you then ask for a Singleton, you will get a new instance in addition to any code using the prior instance.

yeah, it can't and the comment points that out. the reason it exists is to release the module's grip on the singleton so that the garbage collector has a chance of collecting of it. it would require intentional design by the user to use it

Reply:
Well, I think that is fair. Just might want to reference count the instances to avoid the double allocation where we are doing stuff in independent go processes.  Maybe?

Reply:
I think Paul has a good argument for reference counting.

If package A & B grabs the 1gb singleton, then A releases. Then a package C would get a new 1gb singleton.

If the release only released when all references are done, then we'd never have a duplicate singleton out there.